rpc/lib/client/ws_client.go

```diff
@@ -348,21 +379,25 @@ func (c *WSClient) readRoutine() {
 		c.wg.Done()
 	}()
 
-	c.conn.SetReadDeadline(time.Now().Add(c.pongWait))
-
 	c.conn.SetPongHandler(func(string) error {
-		c.conn.SetReadDeadline(time.Now().Add(c.pongWait))
+		// gather latency stats
 		c.mtx.RLock()
-		c.PingPongLatencyTimer.UpdateSince(c.sentLastPingAt)
+		t := c.sentLastPingAt
 		c.mtx.RUnlock()
+		c.PingPongLatencyTimer.UpdateSince(t)
+
 		c.Logger.Debug("got pong")
 		return nil
 	})
 
 	for {
+		// reset deadline for every message type (control or data)
+		if c.readWait > 0 {
+			c.conn.SetReadDeadline(time.Now().Add(c.readWait))
```

Comment:
I thought we didn't want this ? Doesn't it mean we always timeout if a read doesnt come in this time, even if we're not expecting one ? and do the pongs come through here?